src/main/java/com/adroll/cantor/HLLCounter.java

```diff
@@ -453,41 +453,48 @@ public static long intersect(List<HLLCounter> hs) {
       return 0;
     }
     for (HLLCounter hll : hs) {
-      if (hll.size() == 0) {
+      if ((hll.getMinHash() == null && hll.size() == 0) || hll.getMinHash().size() == 0) {
         return 0;
       }
     }
-    TreeSet<Long> all = new TreeSet<Long>();
     int mink = Integer.MAX_VALUE;
     int maxs = Integer.MIN_VALUE;
+    ArrayList<Long[]> minhash_arrays = new ArrayList();
     for(HLLCounter h : hs) {
       if(h.isIntersectable()) {
-        all.addAll(h.getMinHash());
+        minhash_arrays.add(h.getMinHash().toArray(new Long[0]));
```

Comment:
I think the walk over the minhash arrays below depends on the arrays being sorted. It might be worth adding a comment that `h.getMinHash().toArray()` returns the min hash array in sorted order.